File: apps/server/src/routes/connections.ts
import { Elysia, t } from 'elysia'
import { requireAuth } from '$src/auth/gate'
import {
  createConnection,
  deleteConnection,
  getConnection,
  getConnectionSecret,
  listConnections,
  setConnectionDatabase,
  updateConnection,
  type ConnectionInput,
} from '$src/store/connections'
import { closeDriver } from '$src/db/registry'
import { getAdapter } from '$src/db/adapters'
import { withTunnel } from '$src/db/ssh/tunnel'
import type { ProviderId } from '$src/providers'
import type { ConnectionTarget } from '$src/db/types'
import type { SshSecret, SshInput } from '$src/store/connections'

const sslMode = t.Union([
  t.Literal('disable'),
  t.Literal('allow'),
  t.Literal('prefer'),
  t.Literal('require'),
  t.Literal('verify-ca'),
  t.Literal('verify-full'),
])

const sshBody = t.Object({
  enabled: t.Boolean(),
  host: t.String(),
  port: t.Integer({ minimum: 1, maximum: 65535 }),
  username: t.String(),
  authMethod: t.Union([t.Literal('password'), t.Literal('key')]),
  password: t.Optional(t.Nullable(t.String())),
  privateKey: t.Optional(t.Nullable(t.String())),
  passphrase: t.Optional(t.Nullable(t.String())),
})

const connectionBody = t.Object({
  name: t.String({ minLength: 1 }),
  provider: t.Optional(t.Literal('postgresql')),
  host: t.String({ minLength: 1 }),
  port: t.Integer({ minimum: 1, maximum: 65535 }),
  database: t.String({ minLength: 1 }),
  username: t.String({ minLength: 1 }),
  password: t.Optional(t.Nullable(t.String())),
  sslMode,
  color: t.Optional(t.Nullable(t.String())),
  readonly: t.Boolean(),
  ssh: t.Optional(t.Nullable(sshBody)),
})

function toInput(b: typeof connectionBody.static): ConnectionInput {
  return { ...b, provider: b.provider ?? 'postgresql', ssh: (b.ssh ?? null) as SshInput | null }
}

/** Run a connection test via the provider adapter, transparently routing through
 *  an SSH tunnel when requested. For an unsaved test, SSH secrets arrive plaintext.
 *  Dialect-agnostic: it never names PostgreSQL — the adapter does the engine work. */
async function testWithOptionalSsh(
  provider: ProviderId,
  target: ConnectionTarget,
  ssh: SshSecret | null,
) {
  const { testConnection } = getAdapter(provider)
  if (!ssh) return testConnection(target)
  try {
    return await withTunnel({ ...ssh, target: { host: target.host, port: target.port } }, (lp) =>
      testConnection({ ...target, host: '127.0.0.1', port: lp }),
    )
  } catch (e) {
    return { error: (e as Error).message }
  }
}

function sshSecretFromBody(s: typeof sshBody.static | null | undefined): SshSecret | null {
  if (!s || !s.enabled) return null
  return {
    host: s.host,
    port: s.port,
    username: s.username,
    authMethod: s.authMethod,
    password: s.password ?? null,
    privateKey: s.privateKey ?? null,
    passphrase: s.passphrase ?? null,
  }
}

export const connectionsRoutes = new Elysia({ prefix: '/connections' })
  .use(requireAuth)
  .get('/', () => listConnections())
  .post('/', ({ body }) => createConnection(toInput(body)), { body: connectionBody })
  .post(
    '/test',
    async ({ body }) =>
      testWithOptionalSsh(
        body.provider ?? 'postgresql',
        {
          host: body.host,
          port: body.port,
          database: body.database,
          username: body.username,
          password: body.password ?? null,
          sslMode: body.sslMode,
        },
        sshSecretFromBody(body.ssh),
      ),
    { body: connectionBody },
  )
  .get(
    '/:id',
    ({ params, status }) => getConnection(params.id) ?? status(404, { error: 'Not found' }),
  )
  .patch(
    '/:id',
    async ({ params, body, status }) => {
      const updated = updateConnection(params.id, toInput(body))
      if (!updated) return status(404, { error: 'Not found' })
      await closeDriver(params.id) // creds may have changed
      return updated
    },
    { body: connectionBody },
  )
  .delete('/:id', async ({ params, status }) => {
    const ok = deleteConnection(params.id)
    if (!ok) return status(404, { error: 'Not found' })
    await closeDriver(params.id)
    return { deleted: true as const }
  })
  // Switch the active database (reconnects the pool to the new database).
  .post(
    '/:id/database',
    async ({ params, body, status }) => {
      const updated = setConnectionDatabase(params.id, body.name)
      if (!updated) return status(404, { error: 'Not found' })
      await closeDriver(params.id)
      return updated
    },
    { body: t.Object({ name: t.String({ minLength: 1 }) }) },
  )
  .post('/:id/test', async ({ params, status }) => {
    const secret = getConnectionSecret(params.id)
    if (!secret) return status(404, { error: 'Not found' })
    return testWithOptionalSsh(
      secret.provider,
      {
        host: secret.host,
        port: secret.port,
        database: secret.database,
        username: secret.username,
        password: secret.password,
        sslMode: secret.sslMode,
      },
      secret.sshSecret,
    )
  })
  .get(
    '/:id/connection-string',
    ({ params, query, status }) => {
      const secret = getConnectionSecret(params.id)
      if (!secret) return status(404, { error: 'Not found' })
      const withPassword = query.withPassword === 'true'
      const password = withPassword ? secret.password : secret.password ? '****' : null
      return {
        connectionString: getAdapter(secret.provider).buildConnectionString(secret, password),
      }
    },
    { query: t.Object({ withPassword: t.Optional(t.String()) }) },
  )

File: apps/server/src/routes/connections.rs
use std::sync::Arc;
use axum::{
    extract::{Path, Query, State},
    response::Json,
    routing::{get, post},
    Router,
};
use serde::Deserialize;

pub use geto_core::services::connections::{
    ConnectionStringResponse, DeleteResponse, SetDatabaseInput, TestResult,
};
use geto_core::store::connections::{Connection, ConnectionInput};
use crate::error::AppError;
use crate::state::AppState;

#[derive(Deserialize, utoipa::IntoParams)]
#[into_params(parameter_in = Query)]
pub struct ConnectionStringQuery {
    #[serde(rename = "withPassword")]
    pub with_password: Option<String>,
}

#[utoipa::path(
    get,
    path = "/api/connections",
    responses(
        (status = 200, description = "List all connections", body = Vec<Connection>)
    )
)]
pub async fn list_handler(
    State(state): State<Arc<AppState>>,
) -> Result<Json<Vec<Connection>>, AppError> {
    let connections = geto_core::services::connections::list_connections(&state.sqlite_pool).await?;
    Ok(Json(connections))
}

#[utoipa::path(
    post,
    path = "/api/connections",
    request_body = ConnectionInput,
    responses(
        (status = 200, description = "Create connection", body = Connection)
    )
)]
pub async fn create_handler(
    State(state): State<Arc<AppState>>,
    Json(input): Json<ConnectionInput>,
) -> Result<Json<Connection>, AppError> {
    let connection = geto_core::services::connections::create_connection(&state.sqlite_pool, &state.cipher, input).await?;
    Ok(Json(connection))
}

#[utoipa::path(
    get,
    path = "/api/connections/{id}",
    params(
        ("id" = String, Path, description = "Connection ID")
    ),
    responses(
        (status = 200, description = "Get connection by ID", body = Connection),
        (status = 404, description = "Connection not found")
    )
)]
pub async fn get_handler(
    State(state): State<Arc<AppState>>,
    Path(id): Path<String>,
) -> Result<Json<Connection>, AppError> {
    match geto_core::services::connections::get_connection(&state.sqlite_pool, &id).await? {
        Some(conn) => Ok(Json(conn)),
        None => Err(AppError::NotFound("Not found".to_string())),
    }
}

#[utoipa::path(
    patch,
    path = "/api/connections/{id}",
    params(
        ("id" = String, Path, description = "Connection ID")
    ),
    request_body = ConnectionInput,
    responses(
        (status = 200, description = "Update connection", body = Connection),
        (status = 404, description = "Connection not found")
    )
)]
pub async fn update_handler(
    State(state): State<Arc<AppState>>,
    Path(id): Path<String>,
    Json(input): Json<ConnectionInput>,
) -> Result<Json<Connection>, AppError> {
    match geto_core::services::connections::update_connection(&state.sqlite_pool, &state.cipher, &id, input).await? {
        Some(conn) => {
            state.registry.close_driver(&id).await;
            Ok(Json(conn))
        }
        None => Err(AppError::NotFound("Not found".to_string())),
    }
}

#[utoipa::path(
    delete,
    path = "/api/connections/{id}",
    params(
        ("id" = String, Path, description = "Connection ID")
    ),
    responses(
        (status = 200, description = "Delete connection", body = DeleteResponse),
        (status = 404, description = "Connection not found")
    )
)]
pub async fn delete_handler(
    State(state): State<Arc<AppState>>,
    Path(id): Path<String>,
) -> Result<Json<DeleteResponse>, AppError> {
    let deleted = geto_core::services::connections::delete_connection(&state.sqlite_pool, &state.registry, &id).await?;
    if deleted {
        Ok(Json(DeleteResponse { deleted: true }))
    } else {
        Err(AppError::NotFound("Not found".to_string()))
    }
}

#[utoipa::path(
    post,
    path = "/api/connections/{id}/database",
    params(
        ("id" = String, Path, description = "Connection ID")
    ),
    request_body = SetDatabaseInput,
    responses(
        (status = 200, description = "Switch default database", body = Connection),
        (status = 404, description = "Connection not found")
    )
)]
pub async fn set_database_handler(
    State(state): State<Arc<AppState>>,
    Path(id): Path<String>,
    Json(body): Json<SetDatabaseInput>,
) -> Result<Json<Connection>, AppError> {
    match geto_core::services::connections::set_connection_database(&state.sqlite_pool, &state.registry, &id, &body.name).await? {
        Some(conn) => Ok(Json(conn)),
        None => Err(AppError::NotFound("Not found".to_string())),
    }
}

#[utoipa::path(
    post,
    path = "/api/connections/test",
    request_body = ConnectionInput,
    responses(
        (status = 200, description = "Test unsaved connection parameters", body = TestResult)
    )
)]
pub async fn test_unsaved_handler(
    Json(input): Json<ConnectionInput>,
) -> Json<TestResult> {
    let res = geto_core::services::connections::test_unsaved_connection(&input).await;
    Json(res)
}

#[utoipa::path(
    post,
    path = "/api/connections/{id}/test",
    params(
        ("id" = String, Path, description = "Connection ID")
    ),
    responses(
        (status = 200, description = "Test saved connection", body = TestResult),
        (status = 404, description = "Connection not found")
    )
)]
pub async fn test_saved_handler(
    State(state): State<Arc<AppState>>,
    Path(id): Path<String>,
) -> Result<Json<TestResult>, AppError> {
    let res = geto_core::services::connections::test_saved_connection(&state.sqlite_pool, &state.cipher, &id).await?;
    Ok(Json(res))
}

#[utoipa::path(
    get,
    path = "/api/connections/{id}/connection-string",
    params(
        ("id" = String, Path, description = "Connection ID"),
        ConnectionStringQuery
    ),
    responses(
        (status = 200, description = "Get formatted connection URL string", body = ConnectionStringResponse),
        (status = 404, description = "Connection not found")
    )
)]
pub async fn connection_string_handler(
    State(state): State<Arc<AppState>>,
    Path(id): Path<String>,
    Query(q): Query<ConnectionStringQuery>,
) -> Result<Json<ConnectionStringResponse>, AppError> {
    let with_pwd = q.with_password.as_deref() == Some("true");
    let conn_str = geto_core::services::connections::get_connection_string(&state.sqlite_pool, &state.cipher, &id, with_pwd).await?;
    Ok(Json(ConnectionStringResponse { connection_string: conn_str }))
}

pub fn router() -> Router<Arc<AppState>> {
    Router::new()
        .route("/connections", get(list_handler).post(create_handler))
        .route("/connections/test", post(test_unsaved_handler))
        .route(
            "/connections/{id}",
            get(get_handler).patch(update_handler).delete(delete_handler),
        )
        .route("/connections/{id}/database", post(set_database_handler))
        .route("/connections/{id}/test", post(test_saved_handler))
        .route(
            "/connections/{id}/connection-string",
            get(connection_string_handler),
        )
}
